clients/js/src/compat.ts

```diff
@@ -628,39 +641,72 @@ class EnvelopeError extends Error {}
 export async function fetchRuntimePublicKey(
   upstream: UpstreamProvider,
 ): Promise<Uint8Array> {
-  if (
-    isEthers6Signer(upstream) ||
-    isEthers6Provider(upstream) ||
-    isEthers5Signer(upstream) ||
-    isEthers5Provider(upstream)
-  ) {
-    const isSigner = isEthers5Signer(upstream) || isEthers6Signer(upstream);
-    const provider = isSigner ? upstream['provider'] : upstream;
-    if (provider && 'send' in provider) {
-      // first opportunistically try `send` from the provider
-      try {
-        const source = provider as {
-          send: (
-            method: string,
-            params: any[] | ((err: any, ok?: any) => void),
-          ) => Promise<any>;
-        };
-        const arg =
-          'engine' in provider
-            ? // eslint-disable-next-line @typescript-eslint/no-unused-vars
-              function (err: any, ok?: any) {
-                return;
-              }
-            : [];
-        const { key } = await source.send(OASIS_CALL_DATA_PUBLIC_KEY, arg);
-        if (key) return arrayify(key);
-      } catch {
-        // don't do anything, move on to try chainId
+  const isEthersSigner = isEthers5Signer(upstream) || isEthers6Signer(upstream);
+  const provider = isEthersSigner ? upstream['provider'] : upstream;
+  if (provider && 'send' in provider) {
+    // first opportunistically try `send` from the provider
+    try {
+      const source = provider as {
+        send: (
+          method: string | { method: string; params: any[] },
+          params?: any[] | ((err: any, ok?: any) => void),
+        ) => Promise<any> | void;
+      };
+
+      // For Truffle, turn a callback into an synchronous call
+      const deferred = defer<any>();
+      const truffle_callback = function (err: any, ok?: any) {
+        if (ok) {
+          deferred.resolve!(ok.result);
+        }
+        deferred.reject!(err);
+        return;
+      };
+
+      let resp;
+      if (
+        !isEthersSigner &&
+        !isEthers5Provider(provider) &&
+        !isEthers6Provider(provider)
```

Comment:
(nit, optional) Maybe we could just say this is checking for Truffle more explicitly.

Reply:
I think there's somehow a third provider type which gets added into the mix which is causing the non-deterministic failures. But that needs another iteration of bug hunting.

Will have a think about another approach to improve reliability after this is merged.